Aim: To predict whether a pesons heart is ok or not ok

Create a webapp where a user uploads image of the picture(we cant use pictures it isnt image processing, So how so we give a commoner to detect his or her heart) and the app gives it an output  


Clue: It is a binary classsification problem either YES or NO, LR is one of the best model for binary classification 

Procedure:
1. Get the data
2. Data preprocessing
3. Split data to test and training data
4. Feed the data to a ML model (Logstic regression model)
5. Test the model

0 denotes the person doesnt have heart disease
1 denotes the person has a heart disease


In [2]:
#import the dependencies or the libraries

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


In [5]:
#data collection and processing

heartData = pd.read_csv("heartdata.csv")
heartData.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0      63    1   3       145   233    1        0      150      0      2.3   
1      37    1   2       130   250    0        1      187      0      3.5   
2      41    0   1       130   204    0        0      172      0      1.4   
3      56    1   1       120   236    0        1      178      0      0.8   
4      57    0   0       120   354    0        1      163      1      0.6   

   slope  ca  thal  target  
0      0   0     1       1  
1      0   0     2       1  
2      2   0     2       1  
3      2   0     2       1  
4      2   0     2       1

In [6]:
#find the number of rows and columns in the dataset
heartData.shape

(303, 14)

In [13]:
#get more information about the data
heartData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age    303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [12]:
#check for missing values
heartData.isnull().sum()

age      0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [14]:
#Getting statistical measures of the data
heartData.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [16]:
#distribution of target variables
heartData['target'].value_counts()

1    165
0    138
Name: target, dtype: int64

165 persons have heart disease, 138 doesnt have a heart disease

Time to split the features and variables

In [18]:
X = heartData.drop(columns='target', axis=1)
Y = heartData['target']

In [20]:
print(X, Y)


     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0        63    1   3       145   233    1        0      150      0      2.3   
1        37    1   2       130   250    0        1      187      0      3.5   
2        41    0   1       130   204    0        0      172      0      1.4   
3        56    1   1       120   236    0        1      178      0      0.8   
4        57    0   0       120   354    0        1      163      1      0.6   
..      ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298      57    0   0       140   241    0        1      123      1      0.2   
299      45    1   3       110   264    0        1      132      0      1.2   
300      68    1   0       144   193    1        1      141      0      3.4   
301      57    1   0       130   131    0        1      115      1      1.2   
302      57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  
0        0   0     1  
1    

In [24]:
#SPLITTING OUR DATA INTO TEST AND TRAINING DATA
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [25]:
print(X.shape, X_train.shape, X_test.shape)

(303, 13) (242, 13) (61, 13)


In [27]:
print(Y.shape, Y_train.shape, Y_test.shape) #Y is only 1 column so no need for shape

(303,) (242,) (61,)


Model training

In [28]:
model=LogisticRegression()

In [29]:
model.fit(X_train, Y_train)

C:\Users\USER\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

Model evaluation

In [30]:
#accuracy on the training data
X_train_prediction= model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [31]:
print('The accuracy score of the training data is: ',training_data_accuracy )

the accuracy score of the training data is:  0.8512396694214877


In [32]:
#accuracy on the test data
X_test_prediction= model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [33]:
print('The accuracy score of the test data is: ',test_data_accuracy )

The accuracy score of the test data is:  0.819672131147541


Since there is a slight difference btw the accuracy score of the test data and trainng data it means our model is working well
if there is a big difference work on your model again

Create a predictive system

In [37]:
input_data=[57,0,0,120,354,0,1,163,1,0.6,2,0,2] #get a random input data from your dataset and remove the last value
#change it to a numpy array for easy shaping
input_data_to_numpy_array = np.asarray(input_data)
#reshape the array since we are predicting for one instance
input_data_reshaped= input_data_to_numpy_array.reshape(1,-1)

prediction = model.predict(input_data_reshaped)
print(prediction)


if (prediction[0]==0):
    print('Heart disease NOT present')
else:
    print('Heart disease DETECTED!')
    

[1]
Heart disease DETECTED!


C:\Users\USER\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
